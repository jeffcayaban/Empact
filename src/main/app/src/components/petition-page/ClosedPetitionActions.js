import React from 'react';
import {Button, MenuItem, OverlayTrigger, Tooltip} from "react-bootstrap";
import {
    generateGeneralConfirmDialog,
    generateSomethingWentWrongDialog,
    generateSuccessDialog
} from "../../utils/Helpers";
import PetitionReport from "./petition-report/PetitionReport";

/**
 * Provides the user with actions to be applied to a closed petition. This includes deleting or editing a petition.
 */

class ClosedPetitionActions extends React.PureComponent {

    constructor(props) {
        super(props);

        this.state = {
            showReport: false,
        };

        this.handleShowPetitionReport.bind(this);
    }

    // Updates the component's state to show or hide the petition report.
    handleShowPetitionReport() {
        this.setState({ showReport: !this.state.showReport });
    }

    // Handles the result of the petition deletion.
    handleDelete(result, deletePetition) {
        if (result.value) {

            // Assumes that the user has confirmed the action to delete the petition.
            deletePetition()
                .then(response => {
                    generateSuccessDialog('Your petition has been deleted!', 'Explore').then(() => {
                        window.location = `/petitions`
                    })
                }).catch(error => generateSomethingWentWrongDialog())
        }
    }

    // Confirms the request to delete the petition.
    handleDeleteRequest() {
        generateGeneralConfirmDialog().then((result) => this.handleDelete(result, this.props.deletePetition))
    }

    // Generates the button to delete the petition.
    generateDeleteBtn(isOpen) {
        const deleteTooltip = <Tooltip id={"delete-petition-tooltip"}>Permanently delete this petition.</Tooltip>;
        const innerElement = isOpen
            ? <MenuItem onClick={this.handleDeleteRequest.bind(this)}>Delete</MenuItem>
            : <Button bsStyle="danger" onClick={this.handleDeleteRequest.bind(this)}>Delete</Button>;

        return (
            <OverlayTrigger placement={"top"} overlay={deleteTooltip}>
                { innerElement }
            </OverlayTrigger>
        )
    }

    render() {
        const { petition } = this.props;
        const { showReport } = this.state;

        const reportTooltip = (
            <Tooltip id={"petition-report-tooltip"}>Generate a report of the petition and its arguments.</Tooltip>
        );

        return (
            <div>
                <div>
                    <OverlayTrigger placement={"top"} overlay={reportTooltip}>
                        <Button style={{ marginRight: '5px' }} onClick={() => this.handleShowPetitionReport()}>
                            View Report
                        </Button>
                    </OverlayTrigger>
                    { this.generateDeleteBtn(false) }
                </div>
                <PetitionReport petition={petition} showReport={showReport}
                                handleShowPetitionReport={() => this.handleShowPetitionReport()} />
            </div>
        );
    }
}

export default ClosedPetitionActions;
