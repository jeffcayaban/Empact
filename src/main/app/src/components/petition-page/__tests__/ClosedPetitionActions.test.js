import React from 'react';
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";

import ClosedPetitionActions from "../ClosedPetitionActions";

describe('ClosedPetitionActions', () => {

    it('should render successfully', () => {
        const component = shallow(<ClosedPetitionActions />);
        expect(toJson(component)).toMatchSnapshot();
    });

    it('should successfully call handleDelete', () => {
        const component = shallow(<ClosedPetitionActions
            deletePetition={() => Promise.resolve()}
        />);

        const instance = component.instance();
        instance.handleDelete({ value: true }, () => Promise.resolve());
        expect(toJson(component)).toMatchSnapshot();
    });


});
