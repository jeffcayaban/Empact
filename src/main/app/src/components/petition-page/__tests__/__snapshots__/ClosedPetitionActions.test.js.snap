// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClosedPetitionActions should render successfully 1`] = `
<div>
  <div>
    <OverlayTrigger
      defaultOverlayShown={false}
      overlay={
        <Tooltip
          bsClass="tooltip"
          id="petition-report-tooltip"
          placement="right"
        >
          Generate a report of the petition and its arguments.
        </Tooltip>
      }
      placement="top"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "5px",
          }
        }
      >
        View Report
      </Button>
    </OverlayTrigger>
    <OverlayTrigger
      defaultOverlayShown={false}
      overlay={
        <Tooltip
          bsClass="tooltip"
          id="delete-petition-tooltip"
          placement="right"
        >
          Permanently delete this petition.
        </Tooltip>
      }
      placement="top"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="danger"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </Button>
    </OverlayTrigger>
  </div>
  <PetitionReport
    handleShowPetitionReport={[Function]}
    showReport={false}
  />
</div>
`;

exports[`ClosedPetitionActions should successfully call handleDelete 1`] = `
<div>
  <div>
    <OverlayTrigger
      defaultOverlayShown={false}
      overlay={
        <Tooltip
          bsClass="tooltip"
          id="petition-report-tooltip"
          placement="right"
        >
          Generate a report of the petition and its arguments.
        </Tooltip>
      }
      placement="top"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "5px",
          }
        }
      >
        View Report
      </Button>
    </OverlayTrigger>
    <OverlayTrigger
      defaultOverlayShown={false}
      overlay={
        <Tooltip
          bsClass="tooltip"
          id="delete-petition-tooltip"
          placement="right"
        >
          Permanently delete this petition.
        </Tooltip>
      }
      placement="top"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="danger"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </Button>
    </OverlayTrigger>
  </div>
  <PetitionReport
    handleShowPetitionReport={[Function]}
    showReport={false}
  />
</div>
`;
